File: python/examples/quickstart.py
"""Quickstart: a cumulative line, a correction, and what the correction cost.

Run:  python examples/quickstart.py
"""

import copy
import json
from pathlib import Path

from fintech_cumulative_ad_line import (
    calculate_cumulative_advance_decline_line,
    correction_impact,
    diff_lines,
    line_deltas,
    rebase_line,
)

FIXTURES = Path(__file__).resolve().parents[1] / "tests" / "fixtures"
FIXTURE = json.loads((FIXTURES / "ad_line_fixtures.json").read_text(encoding="utf-8"))
CONTRACT, EVENTS = FIXTURE["contract"], FIXTURE["base_events"]
CORRECTION = FIXTURE["synthetic_correction"]

EARLY = "2026-01-09T23:00:00Z"  # before the correction was published
LATE = "2026-01-11T00:00:00Z"   # after it was published

# 1) The line as first published.
original = calculate_cumulative_advance_decline_line(CONTRACT, EVENTS, LATE)
print(f"original ({original['status']}):")
print(f"  net  {[p['net_advances'] for p in original['points']]}")
print(f"  line {[p['cumulative_line'] for p in original['points']]}  final={original['final_value']}")

# 2) A correction to session 2 arrives. ONE session changes; every later level moves.
archive = copy.deepcopy(EVENTS) + [copy.deepcopy(CORRECTION)]
corrected = calculate_cumulative_advance_decline_line(CONTRACT, archive, LATE)
print(f"\nafter correction to {CORRECTION['session_date']}:")
print(f"  net  {[p['net_advances'] for p in corrected['points']]}")
print(f"  line {[p['cumulative_line'] for p in corrected['points']]}  final={corrected['final_value']}")
print(f"  recompute from: {corrected['recompute_from_session']}")

# 3) The diff names it as a restatement, not a market move.
diff = diff_lines(original, corrected)
print(f"\ndiff: {diff['shape']} of {diff['tail_offset']:+d} from {diff['first_divergent_session']} "
      f"({diff['changed_session_count']} levels moved, {diff['reasons']})")

# 4) The same archive read at two knowledge cutoffs — both causal.
impact = correction_impact(CONTRACT, archive, EARLY, LATE)
print(f"\nas of {EARLY}: final={impact['before_final_value']} "
      f"(ignored {impact['ignored_future_event_count']} future event)")
print(f"as of {LATE}: final={impact['after_final_value']}")
print("  -> the line 'moved' because data landed, not because the market did")

# 5) The level is a bookkeeping choice; the deltas are the measurement.
rebased = rebase_line(original, 1000)
print(f"\nrebased to 1000: {[p['cumulative_line'] for p in rebased['points']]}")
print(f"deltas unchanged: {line_deltas(rebased) == line_deltas(original)}")

# 6) A break in the chain drops the suffix rather than summing across the hole.
gapped = [e for e in copy.deepcopy(EVENTS) if e["session_sequence"] != 3]
for index, event in enumerate(gapped, start=1):
    event["ingest_sequence"] = index
broken = calculate_cumulative_advance_decline_line(CONTRACT, gapped, LATE)
print(f"\nsession 3 missing -> status={broken['status']} points={broken['points']} "
      f"final={broken['final_value']}")
print(f"  diagnostic prefix only: sessions {[p['session_sequence'] for p in broken['causal_prefix_diagnostic']]}")

File: typescript/examples/quickstart.ts
/**
 * Quickstart: a cumulative line, a correction, and what the correction cost.
 *
 * Run:  node --experimental-strip-types examples/quickstart.ts
 */

import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";

import {
  calculateCumulativeAdvanceDeclineLine,
  type BreadthEvent,
  type BreadthLineResult,
  type SeriesContract,
} from "../src/cumulativeAdvanceDeclineLine.ts";
import { correctionImpact, diffLines, lineDeltas, rebaseLine } from "../src/propagation.ts";

const FIXTURE = JSON.parse(
  readFileSync(fileURLToPath(new URL("../test/fixtures/ad_line_fixtures.json", import.meta.url)), "utf8"),
);
const CONTRACT: SeriesContract = FIXTURE.contract;
const EVENTS: BreadthEvent[] = FIXTURE.base_events;
const CORRECTION: BreadthEvent = FIXTURE.synthetic_correction;

const EARLY = "2026-01-09T23:00:00Z"; // before the correction was published
const LATE = "2026-01-11T00:00:00Z"; // after it was published

const nets = (r: BreadthLineResult) => r.points.map((p) => p.net_advances);
const lines = (r: BreadthLineResult) => r.points.map((p) => p.cumulative_line);

// 1) The line as first published.
const original = calculateCumulativeAdvanceDeclineLine(CONTRACT, EVENTS, LATE);
console.log(`original (${original.status}):`);
console.log(`  net  [${nets(original)}]`);
console.log(`  line [${lines(original)}]  final=${original.final_value}`);

// 2) A correction to session 2 arrives. ONE session changes; every later level moves.
const archive = [...structuredClone(EVENTS), structuredClone(CORRECTION)];
const corrected = calculateCumulativeAdvanceDeclineLine(CONTRACT, archive, LATE);
console.log(`\nafter correction to ${CORRECTION.session_date}:`);
console.log(`  net  [${nets(corrected)}]`);
console.log(`  line [${lines(corrected)}]  final=${corrected.final_value}`);
console.log(`  recompute from: ${corrected.recompute_from_session}`);

// 3) The diff names it as a restatement, not a market move.
const diff = diffLines(original, corrected);
console.log(
  `\ndiff: ${diff.shape} of ${diff.tail_offset! > 0 ? "+" : ""}${diff.tail_offset} from ` +
    `${diff.first_divergent_session} (${diff.changed_session_count} levels moved, [${diff.reasons}])`,
);

// 4) The same archive read at two knowledge cutoffs — both causal.
const impact = correctionImpact(CONTRACT, archive, EARLY, LATE);
console.log(
  `\nas of ${EARLY}: final=${impact.before_final_value} ` +
    `(ignored ${impact.ignored_future_event_count} future event)`,
);
console.log(`as of ${LATE}: final=${impact.after_final_value}`);
console.log("  -> the line 'moved' because data landed, not because the market did");

// 5) The level is a bookkeeping choice; the deltas are the measurement.
const rebased = rebaseLine(original, 1000);
console.log(`\nrebased to 1000: [${lines(rebased)}]`);
console.log(
  `deltas unchanged: ${JSON.stringify(lineDeltas(rebased)) === JSON.stringify(lineDeltas(original))}`,
);

// 6) A break in the chain drops the suffix rather than summing across the hole.
const gapped = structuredClone(EVENTS)
  .filter((event) => event.session_sequence !== 3)
  .map((event, index) => ({ ...event, ingest_sequence: index + 1 }));
const broken = calculateCumulativeAdvanceDeclineLine(CONTRACT, gapped, LATE);
console.log(
  `\nsession 3 missing -> status=${broken.status} points=[${broken.points}] final=${broken.final_value}`,
);
console.log(
  `  diagnostic prefix only: sessions [${broken.causal_prefix_diagnostic.map((p) => p.session_sequence)}]`,
);
